hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java

```diff
@@ -1329,6 +1333,113 @@ public void deleteUnreferenced(Container container, long localID)
     }
   }
 
+  @Override
+  public ContainerCommandResponseProto streamDataReadOnly(
+      ContainerCommandRequestProto request, KeyValueContainer kvContainer,
+      DispatcherContext dispatcherContext,
+      StreamObserver<ContainerCommandResponseProto> streamObserver) {
+    ContainerCommandResponseProto responseProto = null;
+    try {
+      if (!request.hasReadBlock()) {
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Malformed Read Block request. trace ID: {}",
+              request.getTraceID());
+        }
+        return malformedRequest(request);
+      }
+      ReadBlockRequestProto readBlock = request.getReadBlock();
+      ChunkBuffer data;
+
+      BlockID blockID = BlockID.getFromProtobuf(
+          readBlock.getBlockID());
+      // This is a new api the block should always be checked.
+      BlockUtils.verifyReplicaIdx(kvContainer, blockID);
+
+      BlockData blockData = blockManager.getBlock(kvContainer, blockID);
+      List<ContainerProtos.ChunkInfo> chunkInfos = blockData.getChunks();
+      long blockOffset = 0;
+      int chunkIndex = -1;
+      for (int i = 0; i < chunkInfos.size(); i++) {
+        blockOffset += chunkInfos.get(i).getLen();
+        if (blockOffset > readBlock.getOffset()) {
+          chunkIndex = i;
+          break;
+        }
+      }
+
+      BlockUtils.verifyBCSId(kvContainer, blockID);
```

Comment:
Can this be grouped with `BlockUtils.verifyReplicaIdx(kvContainer, blockID);`?
This might catch any verification failures early and also improve readability

Reply:
done